### PR TITLE
Step duration map configuration parameter ported from the POA Network fork

### PR DESCRIPTION
**Pull request by vkomenda**
The original PR: https://github.com/poanetwork/parity-ethereum/pull/152.

Summary: Instead of a single step duration, this PR allows specifying time intervals with given step durations.

### CODE REVIEW

File: ethcore/src/engines/authority_round/mod.rs [hunk 1]
```diff
@@ -97,13 +99,26 @@ const U16_MAX: usize = ::std::u16::MAX as usize;
...
-			warn!(target: "engine", "step_duration is too high ({}), setting it to {}", step_duration_usize, U16_MAX);
-		}
+		let map_step_duration = |u: ethjson::uint::Uint| {
+			let mut step_duration_usize: usize = u.into();
+			if step_duration_usize == 0 {
+				panic!("AuthorityRoundParams: step duration cannot be 0");
+			}
+			if step_duration_usize > U16_MAX {
+				step_duration_usize = U16_MAX;
+				warn!(target: "engine", "step duration is too high ({}), setting it to {}", step_duration_usize, U16_MAX);
```
Comment (ordian) [inline ethcore/src/engines/authority_round/mod.rs]: not introduced in this PR, but we should probably swap this line with the line above, otherwise we will print `U16_MAX` two times.

File: ethcore/src/engines/authority_round/mod.rs [hunk 2]
```diff
@@ -97,13 +99,26 @@ const U16_MAX: usize = ::std::u16::MAX as usize;
...
+			}
+			step_duration_usize as u16
+		};
+		let step_durations: BTreeMap<u64, u16> = match p.step_duration {
+			StepDuration::Single(u) => iter::once((0, map_step_duration(u))).collect(),
+			StepDuration::Transitions(tr) => {
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: Do we verify there is at least one transition?

File: ethcore/src/engines/authority_round/mod.rs [hunk 3]
```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		let next_step = self.load().checked_add(1)?;
+		for (time, dur) in self.durations.iter().skip(1) {
+			let step_diff = time.checked_add(prev_dur)?.checked_sub(1)?.checked_sub(prev_time)?.checked_div(prev_dur)?;
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: Can't we precompute timestamp->step conversion and store them in the `BTreeMap` instead of calculating it every time?

File: ethcore/src/engines/authority_round/mod.rs [hunk 4]
```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
+
 	fn duration_remaining(&self) -> Duration {
+		self.opt_duration_remaining().unwrap_or_else(|| {
+			let ctr = self.load();
+			error!(target: "engine", "Step counter under- or overflow: {}, aborting", ctr);
+			panic!("step counter under- or overflow: {}", ctr)
+		})
+	}
+
+	fn opt_duration_remaining(&self) -> Option<Duration> {
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: Would be good to have a comment of what (and how) we actually try to achieve in this function. It's not super clear to me. I suspect we want to calculate the remaining duration of current step taking into account potential transitions.

File: ethcore/src/engines/authority_round/mod.rs [hunk 5]
```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
-			None => {
-				let ctr = self.load();
-				error!(target: "engine", "Step counter is too high: {}, aborting", ctr);
-				panic!("step counter is too high: {}", ctr)
-			},
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		let next_step = self.load().checked_add(1)?;
+		for (time, dur) in self.durations.iter().skip(1) {
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: Is this loop supposed to find the transition timestamp and duration of current step? If so I think it could be greately simplified (and optimized) by:
1. Converting timestamps into step transitions + keeping the timestamps
2. Bin searching in the step transitions to find current which transition we fall into and just use the pre-computed starting timestamp.

Instead of bin-searching I think we should rather start from the end, as that's a most likely case that we will fall into.
So to summarize, we need:
`Vec<(StepTransition, TimestampTransition, StepDuration)>`
Then you start from the end and check if `next_step > StepTransition` if it is you use `TimestampTransition + (next_step - StepTransition) * StepDuration` to compute the expected timestamp of the next step.

File: ethcore/src/engines/authority_round/mod.rs [hunk 6]
```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
+			if step >= next_step {
+				break;
+			}
+			prev_step = step;
+			prev_time = step_diff.checked_mul(prev_dur)?.checked_add(prev_time)?;
+			prev_dur = u64::from(*dur);
+		}
+		let time = prev_time.checked_add(next_step.checked_sub(prev_step)?.checked_mul(prev_dur)?)?;
+		let step_end = Duration::from_secs(time);
+		if step_end > now {
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: You can use `saturating_sub` instead of this `if-else`

File: ethcore/src/engines/authority_round/mod.rs [hunk 7]
```diff
@@ -125,54 +140,96 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
+			}
 		}
 	}
 
+	fn opt_calibrate(&self) -> Option<()> {
+		let now = unix_now().as_secs();
+		let mut prev_dur = u64::from(self.durations[&0]);
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		for (time, dur) in self.durations.range(..now).skip(1) {
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: This logic seems convoluted and duplicated, please at least refactor it to a single place.

File: ethcore/src/engines/authority_round/mod.rs [hunk 8]
```diff
@@ -679,23 +736,25 @@ impl<'a, A: ?Sized, B> Deref for CowLike<'a, A, B> where B: AsRef<A> {
 impl AuthorityRound {
 	/// Create a new instance of AuthorityRound engine.
 	pub fn new(our_params: AuthorityRoundParams, machine: Machine) -> Result<Arc<Self>, Error> {
-		if our_params.step_duration == 0 {
-			error!(target: "engine", "Authority Round step duration can't be zero, aborting");
-			panic!("authority_round: step duration can't be zero")
+		if !our_params.step_durations.contains_key(&0) {
+			error!(target: "engine", "Authority Round step 0 duration is undefined, aborting");
+			panic!("authority_round: step 0 duration is undefined")
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: Why do we panic instead of returning a proper error?

File: ethcore/src/engines/authority_round/mod.rs [hunk 9]
```diff
@@ -59,12 +60,13 @@ mod finality;
...
-	/// Time to wait before next block or authority switching,
-	/// in seconds.
+	/// A map defining intervals of blocks with the given times (in seconds) to wait before next
+	/// block or authority switching. The keys in the map are timestamps of starting blocks of those
+	/// periods. The entry at `0` should be defined.
 	///
-	/// Deliberately typed as u16 as too high of a value leads
-	/// to slow block issuance.
-	pub step_duration: u16,
+	/// Wait times (durations) are deliberately typed as `u16` since larger values lead to slow
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: We can still store `u64` to avoid casting in the functions that use it and just validate when creating the params or the engine.

File: ethcore/src/engines/authority_round/mod.rs [hunk 10]
```diff
@@ -2387,7 +2448,7 @@ mod tests {
 		let (_spec, tap, accounts) = setup_empty_steps();
 		let engine = build_aura(|p| {
 			p.validators = Box::new(SimpleList::new(accounts.clone()));
-			p.step_duration = 4;
+			p.step_durations = [(0, 4)].to_vec().into_iter().collect();
```
Comment (tomusdrw) [inline ethcore/src/engines/authority_round/mod.rs]: We need more tests for the actual transitions.

File: ethcore/src/engines/authority_round/mod.rs (hunk 9 above)
Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: Yes, we are already checking that the durations are in the range when receiving `AuthorityRoundParams`. I don't mind storing `u64` in the map.

File: ethcore/src/engines/authority_round/mod.rs [hunk 11]
```diff
@@ -125,54 +148,87 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
+		let (mut prev_dur, mut prev_step, mut prev_time) =
+			self.durations.first().expect("durations cannot be empty");
+		for (step, time, dur) in self.durations.iter().skip(1) {
+ 			if *step >= next_step {
+				break;
+			}
+			prev_step = *step;
+			prev_time = *time;
+			prev_dur = *dur;
+		}
```
Comment (afck) [inline ethcore/src/engines/authority_round/mod.rs]: Maybe simpler:
```rust
let (prev_dur, prev_step, prev_time) = self.durations.iter()
    .take_while(|(_, step, _)| step < next_step)
    .last()
    .expect("durations cannot be empty")
    .clone();
```
(Also in `opt_calibrate`.)

File: ethcore/src/engines/authority_round/mod.rs [hunk 12]
```diff
@@ -125,54 +148,87 @@ impl From<ethjson::spec::AuthorityRoundParams> for AuthorityRoundParams {
...
 		}
 	}
 
+	/// Calibrates the AuRa step number given according to the current time.
+	fn opt_calibrate(&self) -> Option<()> {
+		let now = unix_now().as_secs();
+		let (mut prev_step, mut prev_time, mut prev_dur) =
+			self.durations.first().expect("durations cannot be empty");
+		for (step, time, dur) in self.durations.iter().skip(1) {
+			if *time >= now {
```
Comment (afck) [inline ethcore/src/engines/authority_round/mod.rs]: Should this actually be strict? If `time == now`, I imagine we consider the new step to have started?

File: ethcore/src/engines/authority_round/mod.rs (hunk 8 above)
Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: It's not a question I can answer because the panic is already there. I can change it to returning an error though.

File: ethcore/src/engines/authority_round/mod.rs (hunk 10 above)
Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: I've added a test. Please let me know if you require more.

File: ethcore/src/engines/authority_round/mod.rs (hunk 12 above)
Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: You are right. `==` works better for `Vec`.

File: ethcore/src/engines/authority_round/mod.rs (hunk 11 above)
Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: Actually, it's `(step, _, _)`, which brings the question: do you think there should be a struct instead of the triple?

Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: Ah, sorry! But then it should also be:
```rust
let (prev_step, prev_dur, prev_time) = …
```
A struct would avoid that kind of confusion, yes, but it's also a bit more verbose. Your call.

Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: It's `(step, time, dur)` and I dropped `prev_` as redundant.

File: ethcore/src/engines/authority_round/mod.rs [hunk 13]
```diff
@@ -906,6 +971,12 @@ impl AuthorityRound {
 	fn address(&self) -> Option<Address> {
 		self.signer.read().as_ref().map(|s| s.address() )
 	}
+
+	/// AuRa step getter for testing purposes.
+	#[allow(unused)]
```
Comment (ordian) [inline ethcore/src/engines/authority_round/mod.rs]: Is it actually used in tests? Why not `#[cfg(test)]`?

File: ethcore/src/engines/authority_round/mod.rs [hunk 14]
```diff
@@ -679,23 +726,41 @@ impl<'a, A: ?Sized, B> Deref for CowLike<'a, A, B> where B: AsRef<A> {
...
+			error!(target: "engine", "Authority Round step 0 duration is undefined, aborting");
+			return Err(Error::Engine(EngineError::Config));
+		}
+		if our_params.step_durations.values().any(|v| *v == 0) {
+			error!(target: "engine", "Authority Round step duration cannot be 0");
+			return Err(Error::Engine(EngineError::Config));
 		}
 		let should_timeout = our_params.start_step.is_none();
-		let initial_step = our_params.start_step.unwrap_or_else(|| (unix_now().as_secs() / (our_params.step_duration as u64)));
+		let initial_step = our_params.start_step.unwrap_or(0);
```
Comment (ordian) [inline ethcore/src/engines/authority_round/mod.rs]: I've noticed the `or_else` value has changed, why is it correct?

File: ethcore/types/src/errors/engine_error.rs [hunk 15]
```diff
@@ -83,6 +85,7 @@ impl fmt::Display for EngineError {
 			CliqueWrongAuthorCheckpoint(ref oob) => format!("Unexpected checkpoint author: {}", oob),
 			CliqueFaultyRecoveredSigners(ref mis) => format!("Faulty recovered signers {:?}", mis),
 			CliqueTooRecentlySigned(ref address) => format!("The signer: {} has signed a block too recently", address),
+			Config => format!("Configuration error"),
```
Comment (ordian) [inline ethcore/types/src/errors/engine_error.rs]: Not sure if this error adds any information, we could we Custom error with a message instead.

File: ethcore/src/engines/authority_round/mod.rs [hunk 16]
```diff
@@ -1615,11 +1688,32 @@ impl Engine for AuthorityRound {
 	}
 }
 
+/// A helper accumulator function mapping a step duration and a step duration transition timestamp
+/// to the corresponding step number and the correct starting second of the step.
+fn next_step_time_dur(
```
Comment (ordian) [inline ethcore/src/engines/authority_round/mod.rs]:
```suggestion
fn next_step_time_duration(
```

File: ethcore/src/engines/authority_round/mod.rs [hunk 17]
```diff
@@ -679,23 +726,41 @@ impl<'a, A: ?Sized, B> Deref for CowLike<'a, A, B> where B: AsRef<A> {
...
-		let initial_step = our_params.start_step.unwrap_or_else(|| (unix_now().as_secs() / (our_params.step_duration as u64)));
+		let initial_step = our_params.start_step.unwrap_or(0);
+
+		let mut durations = Vec::new();
+		let mut prev_step = 0u64;
+		let mut prev_time = 0u64;
+		let mut prev_dur = our_params.step_durations[&0];
+		durations.push((prev_step, prev_time, prev_dur));
+		for (time, dur) in our_params.step_durations.iter().skip(1) {
+			let (step, time) = next_step_time_dur(prev_step, prev_time, prev_dur, *time)
```
Comment (ordian) [inline ethcore/src/engines/authority_round/mod.rs]:
```suggestion
			let (step, time) = next_step_time_duration(prev_step, prev_time, prev_dur, *time)
```

File: ethcore/src/engines/authority_round/mod.rs (hunk 14 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: The computation of the initial step is more complicated now, so we call `calibrate()` below instead, to avoid duplicating that code.

File: ethcore/src/engines/authority_round/mod.rs (hunk 13 above)
Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: Right, it's dead code and will be removed.

**Comment by vkomenda** [comment]
Are you happy with the way comments have been addressed? Is there anything else that could be improved?

File: json/src/spec/step_duration.rs [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Step duration configuration parameter
+
+use std::collections::BTreeMap;
+use uint::Uint;
+
+/// Step duration can be specified either as a `Uint` (in seconds), in which case it will be
+/// constant, or as a list of pairs consisting of a block number and a duration, in which case the
```
Comment (ordian) [inline json/src/spec/step_duration.rs]:
```suggestion
/// constant, or as a list of pairs consisting of a timestamp (Unit time) and a duration, in which case the
```

File: ethcore/src/engines/authority_round/mod.rs [hunk 19]
```diff
@@ -57,14 +58,19 @@ use unexpected::{Mismatch, OutOfBounds};
 
 mod finality;
 
+type TransitionStep = u64;
+type TransitionTimestamp = u64;
+type StepDuration = u64;
```
Comment (ordian) [inline ethcore/src/engines/authority_round/mod.rs]: was it changed to `u64` to avoid casts?

Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: We have triples of `u64` which needed some explanation in https://github.com/paritytech/parity-ethereum/blob/7863f8381a3790a3e62c605b6fc27c6c6b7fc5fe/ethcore/src/engines/authority_round/mod.rs#L157.

@ordian Would you prefer a struct type instead of plain triples?

Comment (ordian) [reply ethcore/src/engines/authority_round/mod.rs]: @vkomenda not sure if the fields would be named like `transition_step, transition_timestamp, step_duration`, it would be any clearer, but I don't have a strong preference here.

Comment (vkomenda) [reply ethcore/src/engines/authority_round/mod.rs]: > @vkomenda not sure if the fields would be named like `transition_step, transition_timestamp, step_duration`, it would be any clearer, but I don't have a strong preference here.

That might actually help to avoid confusing different fields when using the triple. I'll add that struct.

File: ethcore/src/engines/authority_round/mod.rs (hunk 6 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: I think `Duration` has no `saturating_sub`. But we could convert `now` to seconds and back, if you prefer.

File: ethcore/src/engines/authority_round/mod.rs (hunk 9 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: But then I wonder whether `u16::MAX` is a good upper bound. This has nothing to do with the number of bits and should just be a threshold above which step durations don't make sense anymore. `u16::MAX` corresponds to about 18 hours. Should we set the limit to 24 hours (86400 seconds)?

File: ethcore/src/engines/authority_round/mod.rs (hunk 2 above)
Comment (afck) [reply ethcore/src/engines/authority_round/mod.rs]: In fact, we should verify that there is a transition for time `0`, i.e. an initial step duration.

**Comment by afck** [comment]
I rebased on `master`.

**Comment by afck** [comment]
Something seems to be wrong with CI.
```
ERROR: Job failed (system failure): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running? (executor_docker.go:739:120s)
```

**Comment by dvdplm** [comment]
@afck Most likely yes and caused by a docker-related change that is missing on this branch. If you merge in master it should be alright. Sorry for the hassle!

**Comment by varasev** [comment]
Seems all checks now have passed. Could we merge it then?
